### PR TITLE
KEYCLOAK-7907 Migration of model tests from old test suite. 

### PR DESCRIPTION
**Pull request by mtrue01**
This is a progress PR,  committting what I have done so far.

### CODE REVIEW

**Comment by pdrozd** [comment]
@mtrue01 Can you update the commit message to include the issue number and the message?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/MultipleRealmsTest.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+    @Test
+    @ModelTest
+    public void testUsers(KeycloakSession session) {
+
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionTU) -> {
+
+
+
+                RealmModel realm1 = sessionTU.realms().createRealm("id1", "realm1");
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/MultipleRealmsTest.java]: Maybe this can be formatted?

Comment (mposolda) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/MultipleRealmsTest.java]: And some of the empty lines can be removed?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+        UserSessionManager localManager = new UserSessionManager(session);
+        for (AuthenticatedClientSessionModel clientSession : userSession.getAuthenticatedClientSessions().values()) {
+            localManager.createOrUpdateOfflineSession(clientSession, userSession);
+            offlineSessions.add(clientSession.getClient().getId());
+        }
+
+        return offlineSessions;
+    }
+
+    private static void resetSession(KeycloakSession session) {
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java]: This method can be probably entirely removed?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+        });*/
+    }
+
+    public static void assertSession(UserSessionModel session, UserModel user, String ipAddress, int started, int lastRefresh, String... clients) {
+        assertEquals(user.getId(), session.getUser().getId());
+        assertEquals(ipAddress, session.getIpAddress());
+        assertEquals(user.getUsername(), session.getLoginUsername());
+        assertEquals("form", session.getAuthMethod());
+        assertTrue(session.isRememberMe());
+        //assertTrue(session.getStarted() >= started - 1 && session.getStarted() <= started + 1);
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java]: These checks should be uncommented

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+    @ModelTest
+    public void testOfflineSessionsCrud(KeycloakSession session) {
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud) -> {
+            // Create some online sessions in infinispan
+            int started = Time.currentTime();
+
+            reloadState(sessionCrud);
+
+            UserSessionModel[] origSessions = createSessions(currentSession);
+            sessionCrud.getTransactionManager().commit();
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java]: As mentioned in the email, these calls to "sessionCrud.getTransactionManager().commit();" should be avoided. And nested transactions should be avoided too.

It's highly possible that even if nested transactions are passing with H2, they will probably fail with MySQL (and some other DBs).

In shortcut, if the old code contained something like this:
```
someMethodCalls1();

resetSession();

someMethodCalls2();
````

You can replace in the new with something like:

````
KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
                someMethodCalls1();
);

KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
                someMethodCalls2();
);
````
However not nested transactions.

Comment (mposolda) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java]: This applies for many places in this class

**Review by mposolda — COMMENTED** [review]
@mtrue01 Thanks for the PR. Added some inline comments.

**Comment by mtrue01** [comment]
I'll take a look and see what I can figure out.  As I said elsewhere, the issue I was running into was the scope of variables.  I will put together an example of what I am running into if I cannot figure it out.

**Comment by mtrue01** [comment]
I have annotated UserSessionProviderOfflineTest.java with my reasoning
behind why I used the nested transactions,   perhaps someone can point me
at a better way.

The thrust of it is that an object does not update properly without commit,
and the commit messes with the transaction manager.   Ending the block as
Marek P. had suggested causes the variable offLineSessions to go out of
scope.  offLineSessions is a hashmap and it is populated and then updated a
few times between line 201 and 220,  so having it go out of scope means
that any changes you've made to it are lost.  My way was to try and keep
the variable in scope and still be able to use the transaction system Marek
P. had put together because it got me around a transaction error.

If there is no commit between the creation of offLineSessions at 201 and
the for loop at 220,  then offlineSessions,size()

All the comments I have added annotating this all begin and end with *****
The first is on line 201:
https://github.com/mtrue01/keycloak/blob/633cd68a0be6b94226367e57ff751e66807cb60c/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java#L201

and the annotations continue over the next 20 lines or so to the conclusion
at line 220:
https://github.com/mtrue01/keycloak/blob/633cd68a0be6b94226367e57ff751e66807cb60c/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java#L220

Any help with this would be greatly appreciated.

--Mark T.

**Comment by mtrue01** [comment]
On Wed, Nov 14, 2018 at 3:41 PM Mark True <mtrue@redhat.com> wrote:

> I have annotated UserSessionProviderOfflineTest.java with my reasoning
> behind why I used the nested transactions,   perhaps someone can point me
> at a better way.
>
> The thrust of it is that an object does not update properly without
> commit, and the commit messes with the transaction manager.   Ending the
> block as Marek P. had suggested causes the variable offLineSessions to go
> out of scope.  offLineSessions is a hashmap and it is populated and then
> updated a few times between line 201 and 220,  so having it go out of scope
> means that any changes you've made to it are lost.  My way was to try and
> keep the variable in scope and still be able to use the transaction system
> Marek P. had put together because it got me around a transaction error.
>
> If there is no commit between the creation of offLineSessions at 201 and
> the for loop at 220,  then offlineSessions,size()
>

This was meant to say:
If there is no commit between the creation of offLineSessions at 201 and
the for loop at 220,  then offlineSessions returns with a size of zero and
the for loop is skipped.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+                // Cleanup
+                realmMgr = new RealmManager(sessionUR);
+                realmMgr.removeRealm(realmMgr.getRealm("foo"));
+            }
+        });
+    }
+
+    @Test
+    @ModelTest
+    public void testExpired(KeycloakSession session) {
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java]: The "testExpired" test was updated in the master due some offlineSessions preloading work. Could you try to rebase and doublecheck that new bits added to the test in the meantime are reflected in the migrated test too?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionPersisterProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.keycloak.testsuite.arquillian.DeploymentTargetModifier.AUTH_SERVER_CURRENT;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class UserSessionPersisterProviderTest extends AbstractTestRealmKeycloakTest {
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionPersisterProviderTest.java]: There were some changes in the test due the offline sessions. Some new methods added (testUpdateTimestamps) and some logic changed in few existing tests. Could you try to make sure that these new tests are migrated as well and it's not lost by migration?

**Review by mposolda — CHANGES_REQUESTED** [review]
@mtrue01 Thanks for all the updates. Looks quite good to me. I left few comments. In summary:
- Some minor formatting issues like more spaces etc (hopefully should be easy for you to fix)
- Some methods changed in the meantime in the UserSessionPersisterProviderTest and UserSessionProviderOfflineTest. Sorry for the complications (it's me who added those), however will be good  to ensure that new tests are not lost during migration
- You commented some assert tests in one of the places (I think it's the check for "started" and "lastSessionRefresh"). Could you uncomment and see if tests passing with it uncommented?
- In one of the tests, you used 1 transaction instead of 3. Maybe it's fine, but will be good to re-test with MySQL. If it's failing, it is probably needed to add back 3 transactions.
- UserSessionPersisterProviderTest can be removed from the old testsuite? It seems you removed most of the tests from the old testsuite, but not this one? Or is it still needed (referenced from some other test?)

**Comment by mposolda** [comment]
Ah, one final note. Once you address the comments, we will probably need to trigger DB matrix job. There is a chance that some tests will fail for some DB as it seems the old testsuite wasn't executed with DB matrix for the long time :( But maybe false alarm, we will see...

**Comment by mabartos** [comment]
This PR is replaced by [PR#5898](https://github.com/keycloak/keycloak/pull/5898)

**Comment by mposolda** [comment]
@mabartos Thanks!

**Pull request closed without merging**